File: src/shared/logger.ts
export interface Logger {
  info(message: string, extra?: Record<string, unknown>): void;
  warn(message: string, extra?: Record<string, unknown>): void;
  error(message: string, extra?: Record<string, unknown>): void;
}

function emit(level: string, message: string, extra?: Record<string, unknown>): void {
  const payload = {
    level,
    message,
    ...extra,
    timestamp: new Date().toISOString()
  };
  console.log(JSON.stringify(payload));
}

export const consoleLogger: Logger = {
  info: (message, extra) => emit("info", message, extra),
  warn: (message, extra) => emit("warn", message, extra),
  error: (message, extra) => emit("error", message, extra)
};

File: internal/shared/logger.go
package shared

import (
	"encoding/json"
	"fmt"
	"time"
)

type Logger interface {
	Info(msg string, extra map[string]interface{})
	Warn(msg string, extra map[string]interface{})
	Error(msg string, extra map[string]interface{})
}

type JSONLogger struct{}

func (l JSONLogger) Info(msg string, extra map[string]interface{})  { emit("info", msg, extra) }
func (l JSONLogger) Warn(msg string, extra map[string]interface{})  { emit("warn", msg, extra) }
func (l JSONLogger) Error(msg string, extra map[string]interface{}) { emit("error", msg, extra) }

func emit(level, message string, extra map[string]interface{}) {
	if extra == nil {
		extra = map[string]interface{}{}
	}
	extra["level"] = level
	extra["message"] = message
	extra["timestamp"] = time.Now().UTC().Format(time.RFC3339)
	b, _ := json.Marshal(extra)
	fmt.Println(string(b))
}
